### PR TITLE
Deploy Oak Hello World example to GCE

### PR DESCRIPTION
**Pull request by conradgrobler**
### Checklist

- [ ] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [ ] I have written tests that cover the code changes.
  - [ ] I have checked that these tests are run by [Cloudbuild](cloudbuild.yaml)
  - [ ] I have updated [documentation](docs/) accordingly.
  - [ ] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.
- [x] Pull request includes prototype/experimental work that is under
      construction.

### CODE REVIEW

File: scripts/delete_example_gcp [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env bash
+
```
Comment (tiziano88) [inline scripts/delete_example_gcp]: Source common script.

File: scripts/deploy_example_gcp [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+bazel run "${bazel_build_flags[@]}" //examples/hello_world/gcp:deploy-oak-gcp-hello-world
+
+# Authenticate as the service account
+gcloud auth activate-service-account --project=$GCP_PROJECT_ID --key-file=$GCP_ACCOUNT_FILE
+
+# Create Oak hello world VM instance and starts the Oak app on it.
+gcloud beta compute instances create oak-hello-world-instance --zone=europe-west2-c \
+--machine-type=n1-standard-1 --tags=oak-hello-world-network --image=oak-gcp-hello-world \
+--boot-disk-size=16GB --boot-disk-type=pd-standard --boot-disk-device-name=oak-hello-world-disk \
+--metadata startup-script="/opt/oak/oak_app_loader_host_loader --enclave_path=/opt/oak/oak_enclave_debug.so --config=/opt/oak/config.bin"
```
Comment (tiziano88) [inline scripts/deploy_example_gcp]: (nit) indent.

Comment (tiziano88) [inline scripts/deploy_example_gcp]: (nit) use single quotes.

File: examples/hello_world/gcp/config.json [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  "builders": [
+    {
+      "type": "googlecompute",
+      "account_file": "{{user `gcp_account_file`}}",
+      "project_id": "{{user `gcp_project_id`}}",
+      "source_image_family": "ubuntu-minimal-1910",
+      "zone": "europe-west2-c",
+      "ssh_username": "ubuntu",
+      "image_name": "oak-gcp-hello-world",
+      "disk_size": 16,
```
Comment (tiziano88) [inline examples/hello_world/gcp/config.json]: 16 GB? Is it possible to specify the unit here?

File: examples/hello_world/gcp/config.textproto [hunk 4]
```diff
@@ -0,0 +1,16 @@
+# proto-file: oak/proto/manager.proto
```
Comment (tiziano88) [inline examples/hello_world/gcp/config.textproto]: This should probably be under the example folder, one level up. cc @ipetr0v

File: scripts/deploy_example_gcp [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+"$SCRIPTS_DIR/build_example" "hello_world"
+
+bazel_build_flags+=(
+    '--config=enc-sim'
+)
+
+# Prerequisites:
+# - Google Cloud SDK
+# - Environment variable GCP_PROJECT_ID: id of the GCP project to use.
```
Comment (tiziano88) [inline scripts/deploy_example_gcp]: So which project should we use? `oak-ci`, or something else? :) Maybe set a default somewhere anyways, it's not like we are going to switch projects very often for now.

File: scripts/deploy_example_gcp [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+#   - Compute Instance Admin (v1)
+#   - Compute Security Admin
+#   - Service Account User 
+
+# Create a minimal ubuntu linux image on the GCP project.
+# The image will be named oak-gcp-hello-world.
+bazel run "${bazel_build_flags[@]}" //examples/hello_world/gcp:deploy-oak-gcp-hello-world
+
+# Authenticate as the service account
+gcloud auth activate-service-account --project=$GCP_PROJECT_ID --key-file=$GCP_ACCOUNT_FILE
```
Comment (tiziano88) [inline scripts/deploy_example_gcp]: Is it not possible to use the default credentials?

File: examples/hello_world/gcp/config.textproto (hunk 4 above)
Comment (ipetr0v) [reply examples/hello_world/gcp/config.textproto]: All config files will be under `examples/${NAME}/config` after https://github.com/project-oak/oak/pull/474 is merged.
Also, I think this config file should be called `gcp_config.textproto` since `config.textproto` will be used by `run_examples`.

File: examples/hello_world/gcp/config.textproto [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+# proto-message: oak.ApplicationConfiguration
```
Comment (ipetr0v) [inline examples/hello_world/gcp/config.textproto]: I think these comments are not necessary, since it's not a `oak/proto/manager.proto` file

File: scripts/deploy_example_gcp (hunk 2 above)
Comment (conradgrobler) [reply scripts/deploy_example_gcp]: Done.

Comment (conradgrobler) [reply scripts/deploy_example_gcp]: Done.

File: scripts/deploy_example_gcp (hunk 5 above)
Comment (conradgrobler) [reply scripts/deploy_example_gcp]: Added a suggestion in the comment (oak-vm-testing). The same environment variable is used by the Packer configuration, and I don't think I should be setting environment variables in the script.

File: examples/hello_world/gcp/BUILD [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+)
+
+exports_files(["config.json"])
+
+assemble_packer(
+    name = "assemble-oak-gcp-hello-world",
+    config = ":config.json",
+    files = {
+        "//oak/server/dev:dev_oak_runner" : "dev_oak_runner",
+        "//examples/hello_world/config:config.bin" : "config.bin",
```
Comment (conradgrobler) [inline examples/hello_world/gcp/BUILD]: As #474 is now merged, this has now been updated to use /examples/hello_world/config:config rather than having a local application config.

File: scripts/deploy_example_gcp [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+# Create Oak hello world VM instance and starts the Oak app on it.
+gcloud beta compute instances create oak-hello-world-instance --zone=europe-west2-c \
```
Comment (tiziano88) [inline scripts/deploy_example_gcp]: It may be more readable to have one flag per line, but up to you.

edit: Have you considered https://cloud.google.com/sdk/gcloud/reference/topic/flags-file for these complex commands? though it may not be a good idea if we rely on variables in this script.

File: scripts/deploy_example_gcp [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+"$SCRIPTS_DIR/build_example" "hello_world"
+
+bazel_build_flags+=(
+    '--config=clang'
+)
+
+# Prerequisites:
+# - Google Cloud SDK
+# - Environment variable GCP_PROJECT_ID: id of the GCP project to use (eg oak-vm-testing).
+# - Environment variable GCP_ACCOUNT_FILE: path to the exported key for the service
```
Comment (tiziano88) [inline scripts/deploy_example_gcp]: Could you add checks for non-emptiness of these variables?

File: scripts/deploy_example_gcp [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+#   account that will be used to connect to GCP.
+# - Service account must be in the following IAM roles:
+#   - Compute Instance Admin (v1)
+#   - Compute Security Admin
+#   - Service Account User 
+
+# Create a minimal ubuntu linux image on the GCP project.
```
Comment (tiziano88) [inline scripts/deploy_example_gcp]:
```suggestion
# Create a minimal Ubuntu linux image on the GCP project.
```

File: scripts/deploy_example_gcp [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+# The image will be named oak-gcp-hello-world.
```
Comment (tiziano88) [inline scripts/deploy_example_gcp]: perhaps make this a `readonly IMAGE_NAME='oak-gcp-hello-world'`

File: scripts/deploy_example_gcp [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+bazel run "${bazel_build_flags[@]}" //examples/hello_world/gcp:deploy-oak-gcp-hello-world
+
+# Authenticate as the service account
+gcloud auth activate-service-account --project=$GCP_PROJECT_ID --key-file=$GCP_ACCOUNT_FILE
```
Comment (tiziano88) [inline scripts/deploy_example_gcp]: (nit) quote and use `${` notation for variables.

https://google.github.io/styleguide/shell.xml?showone=Variable_expansion#Variable_expansion

File: scripts/deploy_example_gcp (hunk 10 above)
Comment (conradgrobler) [reply scripts/deploy_example_gcp]: Added it to gcp_common script.

File: scripts/deploy_example_gcp (hunk 12 above)
Comment (conradgrobler) [reply scripts/deploy_example_gcp]: Added multiple variables to gcp_common.

File: scripts/deploy_example_gcp [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+#   - Compute Security Admin
+#   - Service Account User 
+
+# Create a minimal Ubuntu linux image on the GCP project.
+# The image will be named oak-gcp-hello-world.
+bazel run "${bazel_build_flags[@]}" "//examples/hello_world/gcp:deploy-${IMAGE_NAME}"
+
+# Authenticate as the service account
+gcloud auth activate-service-account \
+  --project="${GCP_PROJECT_ID}" \
```
Comment (conradgrobler) [inline scripts/deploy_example_gcp]: Quoted variables as per style guide.

File: scripts/deploy_example_gcp [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+  --key-file="${GCP_ACCOUNT_FILE}"
+
+# Create Oak hello world VM instance and starts the Oak app on it.
+gcloud beta compute instances create "${INSTANCE_NAME}" \
+  --zone="${ZONE}" \
```
Comment (conradgrobler) [inline scripts/deploy_example_gcp]: Moved to one flag per line for clarity. I don't think a flags file will improve clarity furhter.

File: scripts/deploy_example_gcp [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+
+readonly SCRIPTS_DIR="$(dirname "$(readlink -f "$0")")"
+source "$SCRIPTS_DIR/common"
+
+source "$SCRIPTS_DIR/gcp_common"
+
+"$SCRIPTS_DIR/build_example" "hello_world"
+
+bazel_build_flags+=(
+    '--config=clang'
```
Comment (tiziano88) [inline scripts/deploy_example_gcp]: Why is this needed? Perhaps add a comment.

File: scripts/gcp_common [hunk 17]
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env bash
+
+readonly IMAGE_NAME='oak-gcp-hello-world'
+readonly DISK_NAME='oak-hello-world-disk'
+readonly INSTANCE_NAME='oak-hello-world-instance'
+readonly NETWORK_TAG='oak-hello-world-network'
+readonly FIREWALL_RULE='oak-hello-world-access'
+readonly ZONE='europe-west2-c'
+
+if [[ -z "${GCP_PROJECT_ID:+''}" ]]; then
```
Comment (tiziano88) [inline scripts/gcp_common]: Just curious, why is the `:+''` part needed?

File: scripts/deploy_example_gcp (hunk 16 above)
Comment (conradgrobler) [reply scripts/deploy_example_gcp]: Turns out it is not needed. Removed.

File: scripts/gcp_common (hunk 17 above)
Comment (tiziano88) [reply scripts/gcp_common]: Doesn't that mean that this check is superfluous then?

Comment (conradgrobler) [reply scripts/gcp_common]: This just gives a better error message, making it clear that it is expecting a non-empty environment variable to be set. Without these checks the error is something like 'Unbound variable GCP_PROJECT_ID'.
